### ISSUE TITLE
Bug: HTTP Feature Testing only runs the FIRST test

### ISSUE THREAD

**Issue by 4mesolutions** [codeigniter4/CodeIgniter4#2393]
**Describe the bug**
HTTP Testing is not working properly.
This bug happens when the controller is locates under folder structure.
e.g.: App/Controllers/Folder/Another/Example.php 

When running PhpUnit tests **only the 1st test will pass**, the other tests will return a ERROR:

`CodeIgniter\Exceptions\PageNotFoundException: Controller or its method is not found: App\Controllers\Folder\Another\::example`

If you change the methods order, the test that previously passed will now return a ERROR if it is not at first position.

If the Example.php controller is located under App/Controllers/Example.php all the tests will run normally.

**CodeIgniter 4 version**
4.0 rc3

**Affected module(s)**
HTTP Feature Testing

**Expected behavior, and steps to reproduce if appropriate**
- Add more methods to Home.php controller.
- Place the Home.php under App/Controllers/Folder/Another/Folder/Home.php
- Test it using the HTTP Feature:
```

<?php namespace App\CI4\Controllers;

use CodeIgniter\Test\FeatureTestCase;


class HomeControllerTest extends FeatureTestCase
{

    public function setUp(): void
    {
        parent::setUp();
    }
    
    public function tearDown(): void
    {
        parent::tearDown();
    }

    public function testSimple()
    {
        $test = defined('APPPATH');
        $this->assertTrue($test);
    }

    public function testShowHomeIndex()
    {

        $result = $this->get('Folder/Another/Folder/Home/Index');            

        $this->assertTrue($result->isOK());
    }

    public function testShowHomePage()
    {

        $result = $this->get('Folder/Another/Folder/Home/Page');                

        $this->assertTrue($result->isOK());
    }

    public function testShowHomeMain()
    {

        $result = $this->get('Folder/Another/Folder/Home/main');        

        $this->assertTrue($result->isOK());
    }
} 
```

**Comment by lonnieezell** [comment]
This doesn't sound like a FeatureTestCase issue - but a namespace issue with your class. If the class file is located at `App/Controllers/Folder/Another/Folder/Home.php`.

I've got an example [over here](https://github.com/codeigniter4projects/playground/blob/develop/tests/feature/HomeTest.php) that shows using multiple tests in a FeatureTest and they all run and work fine.

**Comment by 4mesolutions** [comment]
Hi sir !

Just tried it again... Didnt work... 

Can you try calling it from :
` $result = $this->get('Folder/Another/Folder/Home/Index');      `

In your example you called it from :

`$response = $this->get('/');` 

It will work fine if the Home controller is routed to '/' ... but not if it is at a complex path location like 'Folder/Another/Folder/Home/Index'

Can you put your home controller inside Folder/Another/Folder/Home and tell me if it works ?

Can you show me you home controller ?

THANK YOU VERY MUCH MY FRIEND ! BTW, liked your patreon videos this month ! Thank you !

**Comment by 4mesolutions** [comment]
Also... The controller would not work with a wrong namespace.... 

:-1:

**Comment by lonnieezell** [comment]
Thanks for the kind words about the vids. Glad you enjoyed them! 

I'll hopefully get a chance to try this tonight. But quick thought - are you using the current develop branch? There were some URL-related fixes that happened since RC3 that might have an impact here.

**Comment by 4mesolutions** [comment]
No ! Will give it a try with the develop branch ! Will post the results very soon ! Thank you for looking this up !

I got stuck with testing because of this ...

=D

**Comment by MGatner** [comment]
@4mesolutions Any update? Were you able to try your tests against the latest from `develop`?

**Comment by michalsn** [comment]
I checked this issue today and it's valid. After some digging, I found the reason and solution.

So, the problem is here: https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/Router/Router.php#L569

In a given scenario the first test is always passing without problems but the next ones are failing. This is because in the first test we don't have `$this->directory` set, so `$segments` array is "filtered" in a while loop. But when it comes to the next test, we already have `$this->directory` defined, so all segments are returned immediately, causing an incorrect path to a controller.

The solution is simple - just reset `$this->directory` after every test. I decided to not use a shared instance of the Router service because I don't see any benefits. I'm not sure if this can cause any issues coz I didn't play with Router before - but hopefully, it's gonna be ok.

**Comment by MGatner** [comment]
Thanks for the thorough assessment - and your description makes a lot of sense. Looking at the PR, I’m a little uncomfortable changing that in the core but we can see what the tests say. I don’t imagine there are scenarios where someone would modify the request before the routing.

**Comment by michalsn** [comment]
Yes, I wasn't sure about this change either but frankly, I had no other idea where to put it. We'll see how it goes.

**Comment by michalsn** [comment]
After fixing one test it looks good. But I don't know why this test was passing before. When I tried to run only this one file `./vendor/bin/phpunit tests/system/Config/ServicesTest.php` without my changes, tests were failing too, so... a mystery.

Well, it's up to you guys. If you see a better solution - go for it. At least we now know where the problem is.

**Comment by lonnieezell** [comment]
The problem with using a non-shared instance of the router is that it makes it impossible to poll the Router later for what controller name or method name was used, and other stuff like that. 

Maybe we could make CodeIgniter's `setDirectory` method public and allow it to actually be set to null, then have the FeatureTestCase reset that?

**Comment by michalsn** [comment]
I made a new pull request with suggested changes.

**Comment by MGatner** [comment]
I was wondering why you closed, but I like this solution. Waiting on tests...

**Comment by MGatner** [comment]
Re: tests failing... I saw this during the refactor. There were tests passing in TravisCI that would fail locally for me, and would even have errors (like unmatched class names). I’m not sure if it was some weird caching thing but it has me a little uncomfortable.

**Comment by michalsn** [comment]
Actually, when I run tests locally some of them are failing, but it has something to do with my environment or configuration I guess.
